src/replay.rs

```diff
@@ -15,107 +20,120 @@ pub mod dogstatsd {
 // TODO currently missing ability to read tagger state from replay file
 // If this is desired, the length can be found as the last 4 bytes of the replay file
 // Only present in version 2 or greater
-#[derive(Debug)]
-pub struct ReplayReader {
-    buf: Bytes,
+pub struct ReplayReader<'a> {
+    reader: Box<dyn std::io::BufRead + 'a>,
     read_all_unixdogstatsdmsg: bool,
+    _buf: BytesMut,
+}
+
+impl<'a> std::fmt::Debug for ReplayReader<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ReplayReader")
+            .field("read_all_unixdogstatsdmsg", &self.read_all_unixdogstatsdmsg)
+            .finish()
+    }
 }
 
-#[derive(Error, Debug, PartialEq)]
+#[derive(Error, Debug)]
 pub enum ReplayReaderError {
     #[error("No dogstatsd replay marker found")]
     NotAReplayFile,
     #[error("Unsupported replay version")]
     UnsupportedReplayVersion(u8),
+    #[error("IO Error")]
+    Io(#[from] io::Error),
+    #[error("Protobuf Decode error")]
+    ProtoDecode(#[from] DecodeError),
+}
+
+/// header must point to at least 8 bytes
+/// first four should be the replay header magic bytes
+/// next u8 should be the dogstatsd version
+/// next 3 bytes are unused
+///
+/// 8 bytes are always consumed.
+pub fn is_replay(mut header: Bytes) -> Result<(), ReplayReaderError> {
+    assert!(header.len() >= 8);
+
+    // todo is there a better way to grab first 4 into slice?
+    // - slice + advance
+    // - clone + take(4) + into_inner
+    let first_four = header.slice(0..4);
+    header.advance(4);
+    if first_four != DATADOG_HEADER {
+        header.advance(4); // consume next 4 bytes for a total of 8
+        return Err(ReplayReaderError::NotAReplayFile);
+    }
+    // Next byte describes the replay version
+    // f0 is bitwise or'd with the file version, so to get the file version, do a bitwise xor
+    let version = header.get_u8() ^ 0xF0;
+
+    if version != 3 {
+        header.advance(3); // consume next 3 bytes per contract
+        return Err(ReplayReaderError::UnsupportedReplayVersion(version));
+    }
+    header.advance(3); // consume next 3 bytes per contract
+    Ok(())
 }
 
-impl ReplayReader {
+
+impl<'a> ReplayReader<'a> {
     pub fn supported_versions() -> &'static [u8] {
         &[3]
     }
     /// read_msg will return the next UnixDogstatsdMsg if it exists
-    pub fn read_msg(&mut self) -> Option<UnixDogstatsdMsg> {
-        if self.buf.remaining() < 4 || self.read_all_unixdogstatsdmsg {
-            return None;
+    /// TODO, may be useful to explicitly return an EOF error
+    /// rather than None
```

Comment:
Personally, I'd return `None`. It's an expected condition and not strictly speaking an "error" to read to the end of the file.